src/QirRuntime/lib/Tracer/README.md

```diff
@@ -0,0 +1,177 @@
+# Resource Tracer Design Document #
+
+The purpose of the Resource Tracer is to provide efficient and flexible way to estimate resources of a quantum program
+ in QIR representation. The estimates are calculated by simulating execution of the program (as opposed to the static
+ analysis).
+
+In addition to the standard QIR runtime functions, the quantum program will have to:
+
+1. convert _all_ used intrinsic operations into one of the _qis_ operations supported by the tracer (see the list below);
+1. (_optional_) provide callbacks for handling of conditional branches on a measurement (if not provided, the estimates
+ would cover only one branch of the execution);
+1. (_optional_) provide callbacks for start/end of quantum operations (if not provided, all operations will be treated
+ as inlined as if the whole program consisted of a single operation);
+1. (_optional_) provide callbacks for global barriers;
+1. (_optional_) provide description of mapping for frame tracking;
+1. (_optional_) provide names of operations for user friendly output (in the form of `tracer-config.hpp|cpp` files that,
+ ideally, can be generated by the Q# compiler from the mappings described in target.qs).
+
+The last provisions  
+
+The Resource Tracer will consist of:
+
+1. the bridge for the `__quantum__qis__*` extension methods;
+2. the native implementation to back the extensions;
+3. the logic for partitioning gates into layers;
+4. the logic for frame tracking;
+5. output of the collected statistics;
+6. (_lower priority_) the scheduling component to optimize depth and/or width of the circuit.
+
+## List of `__quantum__qis__*` methods ##
+
+___WIP___
+
+| Signature                                             | Description                                                  |
+| :---------------------------------------------------- | :----------------------------------------------------------- |
+| `void __quantum__qis__inject_global_barrier(i32 %id, i32 %duration)` | Function to insert a global barrier. It will be inserted into QIR based on a user defined intrinsic. See [Layering](#layering) section for details. |

```

Comment:
> Function to insert a global barrier.

(I'm a newbie and my questions may be not applicable to your readers)
I would be very pleased if somewhere in the beginning there was some sort of a glossary telling a few words about what the _barrier_ is, how it is different from the _global barrier_, and what it means _to insert a barrier_.
> **It** will be inserted ...

The _function call_ will be inserted or _the global barrier_ will be inserted?
(for my own technical writing I try to avoid pronouns (it, he, me, they, there) and use actual nouns or expressions. At times it results in a repeated words, but technically is clear during the first-most reading, and helps to avoid questions and uncertainty)

Reply:
I moved this list to be the last section of the readme.